### PR TITLE
Fixes #35143 - Add a smart proxy update button

### PR DESCRIPTION
**Pull request by adamruzicka**
Requires:
- https://github.com/theforeman/foreman/pull/9290

### CODE REVIEW

File: app/views/foreman_ansible/job_templates/capsule_upgrade_-_ansible_default.erb [hunk 1]
```diff
@@ -71,3 +74,33 @@ feature: ansible_run_capsule_upgrade
         - name: satellite-maintain upgrade return code is non-zero
           fail:
             msg: "Failed! Capsule server upgrade failed. See /var/log/foreman-installer/capsule.log in the Capsule server for more information"
+<% else %>
```
Comment (stejskalleos) [inline app/views/foreman_ansible/job_templates/capsule_upgrade_-_ansible_default.erb:77]: <% else -%>

**Review by stejskalleos — CHANGES_REQUESTED** [review]
I tried to run playbook on `3.4` version and got following error:

```
TASK [Update foreman repositories] *********************************************
fatal: [proxy34.example.com]: FAILED! => {"msg": "The task includes an option with an undefined variable. The error was: 'ansible_distribution_major_verison' is undefined\n\nThe error appears to be in '/tmp/d20220929-12467-ayq1wo/project/playbook.yml': line 15, column 7, but may\nbe elsewhere in the file depending on the exact syntax problem.\n\nThe offending line appears to be:\n\n\n    - name: Update foreman repositories\n      ^ here\n"}
```

```bash
ansible --version

ansible [core 2.12.2]
  config file = /etc/ansible/ansible.cfg
  configured module search path = ['/root/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python3.8/site-packages/ansible
  ansible collection location = /root/.ansible/collections:/usr/share/ansible/collections
  executable location = /usr/bin/ansible
  python version = 3.8.12 (default, Sep 16 2021, 10:46:05) [GCC 8.5.0 20210514 (Red Hat 8.5.0-3)]
  jinja version = 2.10.3
  libyaml = True
```

File: app/helpers/foreman_ansible/smart_proxies_helper.rb [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+      hosts = proxy.smart_proxy_hosts
+
+      return if !can_schedule_jobs? ||
+                hosts.empty? ||
+                !hosts.all? { |host| can_execute_on_host?(host) }
+
+      version = proxy.statuses[:version].version
+
+      # TODO: Extract this into a service in Foreman
+      ::SmartProxiesController.allocate.send(:versions_mismatched?, version)
```
[the remark this answers is not in the document]
Comment (stejskalleos) [reply app/helpers/foreman_ansible/smart_proxies_helper.rb:13]: IMO we should display the update button only if foreman version > proxy version, it doesn't make sense to display it in opposite case.

**Comment by stejskalleos** [comment]
I hit another issue: I have two proxies, one of them is not available. If you try to access `/smart_proxies`, you end up in never ending loop.

**Comment by adamruzicka** [comment]
> I hit another issue: I have two proxies, one of them is not available. If you try to access `/smart_proxies`, you end up in never ending loop.

I'm unable to reproduce that.

File: app/helpers/foreman_ansible/smart_proxies_helper.rb [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+    def can_update_proxy?(proxy)
+      hosts = proxy.smart_proxy_hosts
+
+      return if !can_schedule_jobs? ||
+                hosts.empty? ||
+                !hosts.all? { |host| can_execute_on_host?(host) }
+
+      version = proxy.statuses[:version].version
+
+      # TODO: Extract this into a service in Foreman
```
[the remark this answers is not in the document]
Comment (adamruzicka) [reply app/helpers/foreman_ansible/smart_proxies_helper.rb]: Not relevant anymore

File: app/helpers/foreman_ansible/smart_proxies_helper.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (adamruzicka) [reply app/helpers/foreman_ansible/smart_proxies_helper.rb]: Done

**Comment by adamruzicka** [comment]
> Job is running for proxy34.example.com

That depends on how the mapping between smart proxies and hosts is set up. The installer usually takes care of this, what does `proxy.infrastructure_host_facets` give you?

> We can have multiple proxies on one host?

Yes and no. Manually you can do almost anything. Barring that, at least the data model allows you to treat a single smart proxy as multiple hosts.

And in general, using capsule in upstream setting is probably wrong.

File: app/views/foreman_ansible/job_templates/capsule_upgrade_-_ansible_default.erb [hunk 4]
```diff
@@ -71,3 +74,33 @@ feature: ansible_run_capsule_upgrade
...
+        manager: auto
+
+    - name: Fail if the target server is a Foreman server
+      fail:
+        msg: "This playbook cannot be executed on a Foreman server. Use only on a Smart Proxy server."
+      when: "'foreman' in ansible_facts.packages"
+
+    - name: Update foreman repositories
+      package:
+        name: https://yum.theforeman.org/releases/{{ foreman_release_version }}/el{{ ansible_distribution_major_version }}/{{ ansible_architecture }}/foreman-release.rpm
```
Comment (stejskalleos) [inline app/views/foreman_ansible/job_templates/capsule_upgrade_-_ansible_default.erb:93]: On RHEL 8 this is failing on:
```
TASK [Update foreman repositories] *********************************************
  21:
fatal: [proxy33.example.com]: FAILED! => {"changed": false, "msg": "Failed to validate GPG signature for foreman-release-3.4.0-1.el8.noarch: Public key for foreman-release3o_lm6dt.rpm is not installed"}
  22:
PLAY RECAP ***
```

However if I run command manually it works without any problems:
```bash
dnf -y install https://yum.theforeman.org/releases/3.4/el8/x86_64/foreman-release.rpm
```

Little rewrite of playbook helped, but I guess it's not the way we wan't to go:
```yaml
    - name: Update foreman repositories
      yum:
        name: https://yum.theforeman.org/releases/{{ foreman_release_version }}/el{{ ansible_distribution_major_version }}/{{ ansible_architecture }}/foreman-release.rpm
        disable_gpg_check: yes
        state: present
```

**Review by stejskalleos — CHANGES_REQUESTED** [review]

File: app/views/foreman_ansible/job_templates/capsule_upgrade_-_ansible_default.erb [hunk 5]
```diff
@@ -44,7 +47,7 @@ feature: ansible_run_capsule_upgrade
                                "--whitelist=#{input('whitelist_options')}"
                              end -%>
       - name: Upgrade Capsule server using satellite-maintain
-        shell: satellite-maintain upgrade run --assumeyes --target-version=<%= input('target_version') %> <%= whitelist_option %>
+        shell: satellite-maintain upgrade run --assumeyes --target-version={{ satellite_version }} <%= whitelist_option %>
```
Comment (stejskalleos) [inline app/views/foreman_ansible/job_templates/capsule_upgrade_-_ansible_default.erb:50]: Note to myself: test it on latest satellite and older capsule as well

**Comment by stejskalleos** [comment]
Failing tests have been fixed in https://github.com/theforeman/foreman_ansible/commit/50c6d566c4cc80ec78184f0f809a19a53eb9aa9d, can you rebase please?

Otherwise :green_apple: **LGTM** :green_apple:, template works correctly, proxy have been successfully updated and everything seems to be working.

Just only one little thing to think about - after clicking on _upgrade_ button job invocation is immediately started,
without any option to check the data before starting the upgrade.
That's fine when users are upgrading proxies with N-1 version, but what about users that have proxy two or more versions behind? For them it can be useful to specify different version than the actual one.
I guess it's edge scenario, definitively not blocking this PR, but I'll bet sooner or later someone will point it out and we'll have to do something with it.

**Comment by adamruzicka** [comment]
Rebased and it should now go through the form instead of being kicked off immediately

**Comment by stejskalleos** [comment]
:green_apple: **LGTM** Works well, proxy upgrades went smoothly, tests are locally passing so let's get this in!

Thanks @adamruzicka !

**Pull request merged**